### PR TITLE
Authentication for Pulse UI

### PR DESCRIPTION
**Pull request by jchristopherinc**

### CODE REVIEW

File: chapter/src/main/java/com/nk/DetailsResorce.java [hunk 1]
```diff
@@ -1,63 +1,80 @@
...
     }
 
+    @POST
+    @Path("adduser")
+    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+     public Response createUser (UserDetails userDetails, @CookieParam("pulseId") int adminId)
+    {
+        if (adminId>0)
+        {
+        repo.addUser(userDetails.getEmail(),adminId);
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsResorce.java]: Don't take it from cookie. Have a cookie table to store cookie id and admin's ID and set the primary key of the entry in cookie. Do a reverse lookup when u get add user request

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 2]
```diff
@@ -56,13 +73,13 @@ public boolean signup (UserDetails a1) {
...
 
-    public  boolean login (UserDetails user) {
-
-        String sql = " SELECT * FROM users WHERE  email= '" + user.email + "'" +  " AND password = crypt('" + user.password + "',password)";
-        System.out.println(sql);
 
-        boolean flag = false;
+    public  int login (UserDetails user) {
+        cookieid = 0;
         Statement st = null;
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: why statement? Why not preparedStatement

**Comment by jchristopherinc** [comment]
Multi tenant concept is missing.

File: chapter/src/main/java/com/nk/DetailsResorce.java [hunk 3]
```diff
@@ -1,63 +1,103 @@
...
-        System.out.println("getting called");
-        String output = "Created customer <a href=\"customers/" +
-                user.getUsername() + "\">" + user.getUsername()
-                + "</a>";
-        String lastVisit = DateFormat.getDateTimeInstance(
-                DateFormat.SHORT, DateFormat.LONG).format(new Date());
-        if(repo.signup(user)){
+        System.out.println(user.getUsername());
+
+        if(repo.signup(user.getUsername(),user.getPassword(),user.getEmail()))
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsResorce.java]: Add input validations / sanitizations. 

What's the guarantee that the input is an email or not?

File: chapter/src/main/java/com/nk/DetailsResorce.java [hunk 4]
```diff
@@ -1,63 +1,103 @@
...
     @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
     public Response userSignup (UserDetails user) {
-        System.out.println("getting called");
-        String output = "Created customer <a href=\"customers/" +
-                user.getUsername() + "\">" + user.getUsername()
-                + "</a>";
-        String lastVisit = DateFormat.getDateTimeInstance(
-                DateFormat.SHORT, DateFormat.LONG).format(new Date());
-        if(repo.signup(user)){
+        System.out.println(user.getUsername());
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsResorce.java]: @anandh619 help her remove `system.out.println`  and move all these to logger for debugging

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 5]
```diff
@@ -75,14 +95,219 @@ public  boolean login (UserDetails user) {
...
+             st.executeUpdate();
+        }catch (SQLException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+
+ }
+  public void  user_app_relation( String name){
+
+    int appid =getAppId(name);
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: Pass it from the previous methods.

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 6]
```diff
@@ -75,14 +95,219 @@ public  boolean login (UserDetails user) {
...
+        }catch (SQLException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+
+
+    }
+    public void addAppInfo (String name){
+
+        String sql = " insert into  apps(appname)  values(?) ";
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: Explore `returning` clause in SQL statement and parse it, to return App object

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 7]
```diff
@@ -75,14 +95,219 @@ public  boolean login (UserDetails user) {
...
+            st.setString(1, name);
+
+             st.executeUpdate();
+        }catch (SQLException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+
+ }
+  public void  user_app_relation( String name){
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: why snake case? use camelCase for Java

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 8]
```diff
@@ -75,14 +95,219 @@ public  boolean login (UserDetails user) {
...
-        return flag ;
+        return cookieid ;
+    }
+
+
+    public  void addUser ( String email,int adminId) {
+
+        String sql = " Insert into invites  values(?,?)";
+        try {
+            if( mailService.sendMail(email)){
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: separate the functionalities. Don't mix email sending in a class where you do DB ops.

File: chapter/src/main/java/com/nk/DetailsResorce.java [hunk 9]
```diff
@@ -1,63 +1,103 @@
...
+     public Response createUser (UserDetails userDetails, @CookieParam("pulseId") int adminId)
+    {
+        if (adminId>0)
         {
-            String output = " customer is  <a href=\"customers/" +
-                user.getEmail() + "\">" + user.getEmail()
-                + "</a>";
-        String lastVisit = DateFormat.getDateTimeInstance(
-                DateFormat.SHORT, DateFormat.LONG).format(new Date());
+        repo.addUser(userDetails.getEmail(),adminId);
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsResorce.java]: Move the email sending part out of DBDAO,

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 10]
```diff
@@ -3,28 +3,40 @@
...
+import java.util.Properties;
 
 public class DetailsRepository {
-    Connection  con = null;
 
+    Connection  con = null;
+    static int cookieid ;
+    Properties props = new Properties();
+    InputStream inputStream;
+    String propFileName = "config.properties";
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: move  it to a constants file.

File: chapter/src/main/java/com/nk/DBInit.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+    public class DBInit {
+
+
+        Properties properties = new Properties();
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DBInit.java]: have access specifier as private

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 12]
```diff
@@ -1,88 +1,184 @@
...
+            st.setString(1, username);
+            st.setString(2, password);
+            st.setString(3, email);
+            rs = st.executeQuery();
+
+            if (rs.next()) {
+                userId = rs.getInt("userid");
+            }
+        } catch (SQLException e) {
+            logger.error("SQL Exception", e);
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: throw a runtime exception and handle it in the place this method is called.

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 13]
```diff
@@ -1,88 +1,184 @@
...
+        }
+        return userId;
+    }
+
 
+    public int login(UserDetails user) {
 
-    public DetailsRepository()  {
+        int userId = 0;
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: needs refactoring

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 14]
```diff
@@ -1,88 +1,184 @@
...
-            // TODO Auto-generated catch block
-            e.printStackTrace();
         }
-        return flag;
     }
 
-    public  boolean login (UserDetails user) {
+    public int addAppInfo(String name) {
+        int appId = 0;
+        String sql = " insert into  apps(appname)  values(?) returning app_id ";
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: 👍

File: chapter/src/main/java/com/nk/DetailsRepository.java [hunk 15]
```diff
@@ -1,88 +1,184 @@
...
+            rs = st.executeQuery();
+            if (rs.next()) {
+                cookieId = rs.getInt("cookie_id");
             }
 
+
+        } catch (SQLException e) {
+            logger.error("SQL Exception", e);
+        }
+        System.out.println("haha" + cookieId);
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsRepository.java]: ?

File: chapter/src/main/java/com/nk/DetailsResource.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        int userId = repo.login(user);
+        String url = "http://localhost:8080/webapi/users/myapps";
+        if (userId > 0) {
+            String cookieValue = String.valueOf(repo.getCookie(userId));
+
+            if (Integer.valueOf(cookieValue) <= 0) {
+                cookieValue = String.valueOf(repo.addCookie(userId));
+            }
+
+            return Response.seeOther(URI.create(url))
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsResource.java]: `seeOther` vs `temporaryRedirect` difference?

File: chapter/src/main/java/com/nk/DetailsResource.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        }
+        else {
+            if (userId == -1)
+                return Response.status(Response.Status.UNAUTHORIZED)
+                        .entity("Email or password is incorrect")
+                        .type("application/json").build();
+        }
+        return Response.status(Response.Status.NOT_FOUND)
+                .entity("email does not exists, please signup")
+                .type("application/json").build();
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsResource.java]: MediaType.APPLICATION_JSON

File: chapter/src/main/java/com/nk/DetailsResource.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @GET
+    @Path("logout")
+    public Response logout(@CookieParam("cookieId") Cookie cookie) {
+
+        if (cookie != null) {
+            NewCookie newCookie = new NewCookie(cookie, null, 0, false);
+            return Response.ok("You have successfully logged out").cookie(newCookie).build();
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/DetailsResource.java]: Remove the cookie entry from DB

File: chapter/src/main/java/com/nk/InputValidation.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+                "A-Z]{2,7}$";
+
+        Pattern pat = Pattern.compile(emailRegex);
+        if (email == null)
+            return false;
+        return pat.matcher(email).matches();
+    }
+    public static boolean isValidPassword(String password)
+    {
+        String passwordRegex = "^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=])(?=\\S+$).{8,}$";
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/InputValidation.java]: see if u can explore https://github.com/nulab/zxcvbn4j

File: chapter/src/main/java/com/nk/InputValidation.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+        Pattern pat = Pattern.compile(passwordRegex );
+        if (password == null)
+            return false;
+        return pat.matcher(password).matches();
+    }
+    public static boolean isValidUsername(String username)
+    {
+        String passwordRegex = "^[a-z0-9_-]{3,15}$";
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/InputValidation.java]: 😠 Please copy paste and change correctly. `passwordRegex`

File: chapter/src/main/java/com/nk/MailService.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+public class MailService {
+
+
+    public boolean sendMail(String invitee_Mail) {
+
+        final String username = "nandini2052@gmail.com"; // enter your mail id
+        final String password = " Akkadevatalu1 ";// enter ur password
+
+        Properties props = new Properties();
+        props.put("mail.smtp.auth", "true");
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/MailService.java]: constants file

File: chapter/src/main/java/com/nk/MailService.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+                new javax.mail.Authenticator() {
+                    protected PasswordAuthentication getPasswordAuthentication() {
+                        return new PasswordAuthentication(username, password);
+                    }
+                });
+
+        try {
+
+            Message message = new MimeMessage(session);
+            message.setFrom(new InternetAddress("nandini2052@gmail.com")); // same email id
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/MailService.java]: Yes, we know its gonna be the same email ID. Please use the constants.

File: chapter/src/main/java/com/nk/MailService.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+            message.setRecipients(Message.RecipientType.TO,
+                    InternetAddress.parse(invitee_Mail));// whome u have to send mails that person id
+            message.setSubject("Testing Subject");
+            message.setText("Dear Mail Crawler,"
+                    + "\n\n No spam to my email, please!");
```
Comment (nandini2052) [inline chapter/src/main/java/com/nk/MailService.java]: ha ha

File: chapter/src/main/resources/db/migration/V9__create_cookies_table.sql [hunk 24]
```diff
@@ -0,0 +1 @@
+create table cookies_list ( cookie_id serial primary key , user_id integer not null references users(userid));
```
Comment (nandini2052) [inline chapter/src/main/resources/db/migration/V9__create_cookies_table.sql:1]: use bigserial. reason being serial is int and bigserial is long

File: chapter/src/main/resources/logback.xml [hunk 25]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration>
+    <appender name="fileAppender" class="ch.qos.logback.core.FileAppender">
```
Comment (nandini2052) [inline chapter/src/main/resources/logback.xml:3]: FileAppender is 👍 would be great to see `RollingFileAppender`

File: chapter/src/main/resources/project.properties [hunk 26]
```diff
@@ -0,0 +1,7 @@
+DATABASE_USER_NAME=nandinik
+DATABASE_PASSWORD=nandinidb
+DRIVER_CLASS_NAME=org.postgresql.Driver
+DATABASE_PORT=5432
+SHARD_ZERO_HOST=localhost
+DATABASE_URL=jdbc:postgresql://localhost:5432/pingdb
```
Comment (nandini2052) [inline chapter/src/main/resources/project.properties]: generate this logic in code.. move pingdb as a property

**Review by nandini2052 — COMMENTED** [review]
Ha ha changes

File: chapter/src/main/java/com/nk/QueryHelper.java [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+public class QueryHelper {
+
+    private final static Logger log = LoggerFactory.getLogger(QueryHelper.class);
+
+    public Connection createConnection(){
```
Comment (jchristopherinc) [inline chapter/src/main/java/com/nk/QueryHelper.java:12]: 👍

**Review by jchristopherinc — COMMENTED** [review]
LGTM

**Pull request merged**